File: runtime/arith.go
package runtime

import (
	"math"
)

// Unm returns (z, true) where z is the value representing -x if x is a number,
// else (NilValue, false).
func Unm(x Value) (Value, bool) {
	switch x.iface.(type) {
	case int64:
		return IntValue(-x.AsInt()), true
	case float64:
		return FloatValue(-x.AsFloat()), true
	default:
		return NilValue, false
	}
}

// Add returns (z, true) where z is the value representing x+y if x and y are
// numbers, else (NilValue, false).
func Add(x, y Value) (Value, bool) {
	switch x.iface.(type) {
	case int64:
		switch y.iface.(type) {
		case int64:
			return IntValue(x.AsInt() + y.AsInt()), true
		case float64:
			return FloatValue(float64(x.AsInt()) + y.AsFloat()), true
		}
	case float64:
		switch y.iface.(type) {
		case int64:
			return FloatValue(x.AsFloat() + float64(y.AsInt())), true
		case float64:
			return FloatValue(x.AsFloat() + y.AsFloat()), true
		}
	}
	return NilValue, false
}

// Sub returns (z, true) where z is the value representing x-y if x and y are
// numbers, else (NilValue, false).
func Sub(x, y Value) (Value, bool) {
	switch x.iface.(type) {
	case int64:
		switch y.iface.(type) {
		case int64:
			return IntValue(x.AsInt() - y.AsInt()), true
		case float64:
			return FloatValue(float64(x.AsInt()) - y.AsFloat()), true
		}
	case float64:
		switch y.iface.(type) {
		case int64:
			return FloatValue(x.AsFloat() - float64(y.AsInt())), true
		case float64:
			return FloatValue(x.AsFloat() - y.AsFloat()), true
		}
	}
	return NilValue, false
}

// Mul returns (z, true) where z is the value representing x*y if x and y are
// numbers, else (NilValue, false).
func Mul(x, y Value) (Value, bool) {
	switch x.iface.(type) {
	case int64:
		switch y.iface.(type) {
		case int64:
			return IntValue(x.AsInt() * y.AsInt()), true
		case float64:
			return FloatValue(float64(x.AsInt()) * y.AsFloat()), true
		}
	case float64:
		switch y.iface.(type) {
		case int64:
			return FloatValue(x.AsFloat() * float64(y.AsInt())), true
		case float64:
			return FloatValue(x.AsFloat() * y.AsFloat()), true
		}
	}
	return NilValue, false
}

// Div returns (z, true) where z is the (float) value representing x/y if x and
// y are numbers, else (NilValue, false).
func Div(x, y Value) (Value, bool) {
	switch x.iface.(type) {
	case int64:
		switch y.iface.(type) {
		case int64:
			return FloatValue(float64(x.AsInt()) / float64(y.AsInt())), true
		case float64:
			return FloatValue(float64(x.AsInt()) / y.AsFloat()), true
		}
	case float64:
		switch y.iface.(type) {
		case int64:
			return FloatValue(x.AsFloat() / float64(y.AsInt())), true
		case float64:
			return FloatValue(x.AsFloat() / y.AsFloat()), true
		}
	}
	return NilValue, false
}

func floordivInt(x, y int64) int64 {
	r := x % y
	q := x / y
	if r != 0 && (r < 0) != (y < 0) {
		q--
	}
	return q
}

func floordivFloat(x, y float64) float64 {
	return math.Floor(x / y)
}

// Div returns (z, true, nil) where z is the (integer) value representing x//y
// if x and y are numbers and y != 0, if y == 0 it returns (NilValue, true,
// div_by_zero_error), else (NilValue, false, nil) if x or y is not a number.
func Idiv(x Value, y Value) (Value, bool, *Error) {
	switch x.iface.(type) {
	case int64:
		switch y.iface.(type) {
		case int64:
			ny := y.AsInt()
			if ny == 0 {
				return NilValue, true, NewErrorS("attempt to divide by zero")
			}
			return IntValue(floordivInt(x.AsInt(), ny)), true, nil
		case float64:
			return FloatValue(floordivFloat(float64(x.AsInt()), y.AsFloat())), true, nil
		}
	case float64:
		switch y.iface.(type) {
		case int64:
			return FloatValue(floordivFloat(x.AsFloat(), float64(y.AsInt()))), true, nil
		case float64:
			return FloatValue(floordivFloat(x.AsFloat(), y.AsFloat())), true, nil
		}
	}
	return NilValue, false, nil
}

func modInt(x, y int64) int64 {
	r := x % y
	if r != 0 && (r < 0) != (y < 0) {
		r += y
	}
	return r
}

func modFloat(x, y float64) float64 {
	r := math.Mod(x, y)
	if r != 0 && (r < 0) != (y < 0) {
		r += y
	}
	return r
}

// Mod returns (z, true, nil) where z is the (integer or float) value
// representing x%y if x and y are numbers and y != 0, if y == 0 it returns
// (NilValue, true, mod_by_zero_error), else (NilValue, false, nil) if x or y is
// not a number.
func Mod(x Value, y Value) (Value, bool, *Error) {
	switch x.iface.(type) {
	case int64:
		switch y.iface.(type) {
		case int64:
			ny := y.AsInt()
			if ny == 0 {
				return NilValue, true, NewErrorS("attempt to perform 'n%0'")
			}
			return IntValue(modInt(x.AsInt(), ny)), true, nil
		case float64:
			return FloatValue(modFloat(float64(x.AsInt()), y.AsFloat())), true, nil
		}
	case float64:
		switch y.iface.(type) {
		case int64:
			return FloatValue(modFloat(x.AsFloat(), float64(y.AsInt()))), true, nil
		case float64:
			return FloatValue(modFloat(x.AsFloat(), y.AsFloat())), true, nil
		}
	}
	return NilValue, false, nil
}

func powFloat(x, y float64) float64 {
	return math.Pow(x, y)
}

// Pow returns (z, true) where z is the (float) value representing x^y if x and
// y are numbers, else (NilValue, false).
func Pow(x, y Value) (Value, bool) {
	var fx, fy float64
	switch x.iface.(type) {
	case int64:
		fx = float64(x.AsInt())
	case float64:
		fx = x.AsFloat()
	default:
		return NilValue, false
	}
	switch y.iface.(type) {
	case int64:
		fy = float64(y.AsInt())
	case float64:
		fy = y.AsFloat()
	default:
		return NilValue, false
	}
	return FloatValue(powFloat(fx, fy)), true
}

func binaryArithFallback(t *Thread, op string, x, y Value) (Value, *Error) {
	res, err, ok := metabin(t, op, x, y)
	if ok {
		return res, err
	}
	return NilValue, BinaryArithmeticError(op[2:], x, y)
}

// BinaryArithmeticError returns an error describing the problem with trying to
// perform x op y.
func BinaryArithmeticError(op string, x, y Value) *Error {
	var wrongVal Value
	switch {
	case numberType(y) != NaN:
		wrongVal = x
	case numberType(x) != NaN:
		wrongVal = y
	default:
		return NewErrorF("attempt to %s a '%s' with a '%s'", op, x.CustomTypeName(), y.CustomTypeName())
	}
	return NewErrorF("attempt to perform arithmetic on a %s value", wrongVal.CustomTypeName())
}

func unaryArithFallback(t *Thread, op string, x Value) (Value, *Error) {
	res, err, ok := metaun(t, op, x)
	if ok {
		return res, err
	}
	return NilValue, UnaryArithmeticError(op[2:], x)
}

// UnaryArithmeticError returns an error describing the problem with trying to
// perform the unary operation op(x).
func UnaryArithmeticError(op string, x Value) *Error {
	return NewErrorF("attempt to %s a '%s'", op, x.CustomTypeName())
}

File: runtime/lua/arith.lua
print(-(1+1))
--> =-2

print(-1.01)
--> =-1.01

print(pcall(function(x) return -x end, {}))
--> ~^false\t

print(1 + 1.2, 2.2 + 1, 3.2 + 1.2)
--> =2.2	3.2	4.4

local function testbin(f, x, y)
    print(not pcall(f, x, y) and not pcall(f, y, x))
end

testbin(function(x, y) return x + y end, 1, {})
--> =true

print(1 - 2, 3.5 - 2, 4 - 1.5, 4.5 - 2.5)
--> =-1	1.5	2.5	2
 
testbin(function(x, y) return x - y end, 1, {})
--> =true

print(1 * 1, 1.5 * 3, 2 * 2.5, 3.0 * 6.0)
--> =1	4.5	5	18

testbin(function(x, y) return x * y end, 1, {})
--> =true

print(4 / 2, 2 / 0.5, 4.0 / 2, 1.5 / 0.5)
--> =2	4	2	3

testbin(function(x, y) return x / y end, 1, {})
--> =true

print(4 // 2, 4.5 // 2, 2 // 0.8, 3.5 // 0.5)
--> =2	2	2	7

testbin(function(x, y) return x // y end, 1, {})
--> =true

print(5 % 2, 3.5 % 2, 2 % 0.5, 3.5 % 0.5)
--> =1	1.5	0	0

print(-3 % 2, -3.0 % 2)
--> =1	1

testbin(function(x, y) return x % y end, 1, {})
--> =true

print(3^2, 9^0.5, 0.5^2, 4.0^1.5)
--> =9	3	0.25	8

testbin(function(x, y) return x^y end, 1, {})
--> =true

testbin(function(x, y) return x^y end, 1, "a")
--> =true

testbin(function(x, y) return x+y end, 1, "12x")
--> =true

print(pcall(function() return 1//0 end))
--> ~false\t.*divide by zero

print(pcall(function() return 1%0 end))
--> ~false\t.*perform 'n%0'

print(0/0 == 0/0)
--> =false

print(1/0 == 1/0)
--> =true

print(-0.0 == 0.0)
--> =true

print(" 2 " * " 70")
--> =140

print(" 2e3  " ^ "  2.0" == 4000000)
--> =true

do
    local n = 100000000000000000000

    print(n == 1e20)
    --> =true

    print(math.type(n))
    --> =float
end

print("0x8.8" + 0)
--> =8.5

print("0x0.4p2" + 0)
--> =1

n=-12
print(n%n)
--> =0

n=math.mininteger
print(n%n)
--> =0

print("0xffff00000000000000000000000000000000001" + 1)
--> =2
